README.md
# Lane Keeping Assist System
Project of University of Applied Science Karlsruhe, Lane Keeping Assist System in Unity using OpenCV.

## Windows Environment (/Windows)

The simulation runs on windows in Unity and streams images of the front center camera and the current velocity of the vehicle towards linux on a TCP/IP stream.

## Linux Environment (/Linux)

This is where the lane detection is done. After detecting the lane and calculating a waypoint,
the pure pursuit algorithm is used to calculate the steer, gas and brake command and these commands
are send back.

### Requirements

OS: Ubuntu 18.04  
Fixed IP Adress: 192.168.206.3  
Ubuntu-Packages: `sudo apt-get install python python-pip`  
pip-Packages: `pip install numpy opencv-python`  


Changelog.txt
### Version 1.5.1 - Release: 12.03.2020
- Fixing RAM bug of Unity
- Removing unused code
- Updating readme files
- Updating test environment
- Adding reverse lights
- Adjusting lights of vehicle

### Version 1.5 - Release: 26.02.2020
- Adding information about connection in GUI
- Calibrating camera
- Adjusting parameters
- Fixing flickering of way point
- Updating Unity project to 2019.3
- Adding main menu
- Adding daytime changing
- Adding fog changing
- Adding scene with lane line errors

### Version 1.4.1 - Release: 19.02.2020
- Adding "@" as delimiter in TCP receive string
- Adding reconnection of unity to TCP port
- Adding TCP time out
- New test circuit map
- Minor bug fixes
- Code cleaning

### Version 1.4.0 - Release: 04.02.2020
- Adding test environment
- Adding unity build
- Adding drag value to vehicle
- Adding draw thickness to parameters
- Fixing error of lane line prediction
- Removing dynamic threshold of acceleration
- Updating READMEs
- Minor bug fixes

### Version 1.3.0 - Release: 21.12.2019
- Adjusting Parameters
- Adding Line Prediction
- Fixing Pure Pursuit
- Fixing Waypoint Generation
- Adding Visualization Parameters

### Version 1.2.0 - Release: 12.12.2019
- Adding quadratic interpolation
- Adjusting Parameters

### Version 1.1.0 - Release: 10.12.2019
- Calculation of Brake and Gas Pedal is now in Pure Pursuit File
- Generating an main file where you can set the parameters
- New structuring of workspace
- New structuring of modules
- Adding Look Ahead ratio

### Version 1.0.0 - Release: 09.12.2019
- Initial Implementation of Lane Detection
- Initial Implementation of Pure Pursuit
- Initial Implementation of TCP Server Script


Linux/Testing/test.py
# ======================================================
# Copyright (C) 2020 repa1030
# This program and the accompanying materials
# are made available under the terms of the MIT license.
# ======================================================

import cv2
import sys
import numpy as np
import os
sys.path.append('../')
from LaneDetector import lane_detector

# Pure Pursuit Settings
look_ahead_ratio = 1.0          # Default: 1.0 (smoothing of steering)
max_look_ahead_distance = 7.0   # Default: 7.0 [m] (if -1, it is calculated automatically)

# CV Settings
use_quadratic_interpol = False  # Default: False (don't use this)
offset_to_clipping_plane = 3.83 # Default: 3.83 [m] (base link to bottom border of image) 
pixel_to_cm = 0.25              # Default: 0.25 (change resolution of line image)
padding_x = 100                 # Default: 150 [cm] (padding on both sides)
padding_y = 700                 # Default: 700 [cm] (padding to forward)

# Draw Settings
draw_line_prediction = True     # Default: True (draw predicted line)
draw_waypoint = True            # Default: True (draw next waypoint on image)
draw_thickness = 3              # Default: 3 (thickness of lines and waypoint)
waypoint_radius = 4             # Default: 4 (radius of the drawn waypoint)
color_right_line = [0, 255, 0]  # Default: [0, 255, 0] [B G R]
color_left_line = [255, 0, 0]   # Default: [255, 0, 0] [B G R]
color_pred_line = [0, 0, 255]   # Default: [0, 0, 255] [B G R]
color_waypoint = [0, 140, 255]  # Default: [0, 140, 255] [B G R]

src_p0 = np.float32([[150, 540], [808, 540], [571, 344], [389, 344]]) # UL, UR, OR, OL
src_p1 = np.float32([[155, 337], [525, 337], [410, 255], [275, 255]]) # UL, UR, OR, OL 
src_p2 = np.float32([[0, 880], [760, 880], [510, 700], [240, 700]]) # UL, UR, OR, OL 
src_p3 = np.float32([[90, 341], [425, 341], [350, 260], [230, 260]]) # UL, UR, OR, OL 
src_p4 = np.float32([[140, 340], [490, 340], [345, 255], [255, 255]]) # UL, UR, OR, OL
src_p5 = np.float32([[105, 341], [365, 341], [335, 270], [260, 270]]) # UL, UR, OR, OL 
src_p6 = np.float32([[150, 2157], [2915, 2157], [2630, 1680], [1500, 1680]]) # UL, UR, OR, OL 

src_points = [['test0', src_p0],
              ['test1', src_p1],
              ['test2', src_p2],
              ['test3', src_p3],
              ['test4', src_p4],
              ['test5', src_p5],
              ['test6', src_p6]]

dst_p = np.float32([[int((0 + padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm)],
                        [int((300 + padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm)], 
                        [int((300 + padding_x) * pixel_to_cm), int((0 + padding_y) * pixel_to_cm)], 
                        [int((0 + padding_x) * pixel_to_cm), int((0 + padding_y) * pixel_to_cm)]])
warp_size = (int((300 + 2 * padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm))

i = 0
img_folder = 'Images'
for file in os.listdir(img_folder):
    file_str = str(file)
    file_path = os.path.join(img_folder, file)
    img = cv2.imread(img_folder + '/' + file_str)
    for obj in src_points:
        if file_str.find(obj[0]) is not -1:
            src_p = obj[1]
            break

    ld = lane_detector.SimpleLaneDetector(src_p, dst_p, warp_size, look_ahead_ratio,
                                            max_look_ahead_distance, pixel_to_cm, 
                                            use_quadratic_interpol, draw_line_prediction,
                                            draw_waypoint, color_right_line, color_left_line, 
                                            color_pred_line, color_waypoint, draw_thickness,
                                            waypoint_radius)
    (output, wp) = ld.detectLaneLines(img, 0)
    warp_mat = cv2.getPerspectiveTransform(src_p, dst_p)
    image = cv2.warpPerspective(img, warp_mat, warp_size)

    print(file_str + '\t' + str(wp))
    cv2.imwrite('Warped/' + file_str + '_warped.jpg', image)
    cv2.imwrite('Results/' + file_str + '_detection.jpg', output)

    i += 1





Linux/TcpServer/tcp_server.py
#!/usr/bin/env python
# ======================================================
# Copyright (C) 2020 repa1030
# This program and the accompanying materials
# are made available under the terms of the MIT license.
# ======================================================
''' 
TCP Server for receiving images from uniy on linux
Requires openCV and numpy for python
'''
import socket
import numpy as np
import cv2
import time

class TcpServer:

    def __init__(self, tcp_ip, tcp_port, buffer_size, time_out):
        self.ip = tcp_ip
        self.port = tcp_port
        self.buf_size = buffer_size
        self.conn = None
        self.addr = None
        self.tcp_time_out = time_out

    def connectToClient(self):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind((self.ip, self.port))
        s.listen(1)
        self.conn, self.addr = s.accept()
        self.conn.settimeout(self.tcp_time_out)

    def closeConnection(self):
        self.conn.close()

    def reconnect(self):
        self.conn.close()
        print ('Reconnecting to Client...')
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind((self.ip, self.port))
        s.listen(1)
        self.conn, self.addr = s.accept()
        self.conn.settimeout(self.tcp_time_out)
        print ('Connection to Client is established.')

    def receiveMessage(self):
        msg = ''
        try:
            # While there is content in buffer
            while True:
                data = self.conn.recv(self.buf_size)
                msg = msg + data
                if len(data) < self.buf_size: break
                data = None
            # Check if buffer is not empty
        except socket.timeout:
            return (None, None, False)
        img = None
        if len(msg) > 0:
            # Decoding jpg image to array and read current velocity
            msg = np.fromstring(msg, np.uint8)
            velo = msg[-2]
            img = cv2.imdecode(msg[:-2], -1)
        if type(img) is np.ndarray:
            return (img, velo, True)
        return (None, None, True)

    def sendMessage(self, m_x, m_y, m_z):
        msg = 'START@' + str(m_x) + ';' + str(m_y) + ';' + str(m_z) + '@END'
        try:
            self.conn.send(msg)
            return True
        except:
            return False


Linux/PurePursuit/pure_pursuit.py
#!/usr/bin/env python
# ======================================================
# Copyright (C) 2020 repa1030
# This program and the accompanying materials
# are made available under the terms of the MIT license.
# ======================================================
'''
Core calculation of pure pursuit algorithm
The algorithm calculates a circle through the
current position of the car and the next waypoint.
This returns a curvature which can be converted to the
single track steering model.
'''

import math
import numpy as np

class PurePursuit:

    # initialization of instance of pure pursuit
    def __init__(self, cmd_velo, vehicle_mass, wheel_rad, motor_torque, 
                    brake_torque, wheel_base, offset):
        self.next_wp = np.zeros(3) # [x1, y1, z1], z1 is not used
        self.current_pos = np.zeros(3) # [x0, y0, z0], z0 is not used
        self.max_radius = 9e10 # max radius of steering
        self.wheel_base = wheel_base # wheel base [m], used for Ackermann
        self.offset = offset # offset from vehicle to clipping plane
        self.wp_velocity = cmd_velo * 1000.0 / (60.0 * 60.0)
        self.max_accel = motor_torque / wheel_rad / vehicle_mass
        self.max_decel = brake_torque / wheel_rad / vehicle_mass
        self.gas = 0.0      # init
        self.brake = 0.0    # init
        self.steer = 0.0    # init

    # this converts the curvature to a steering angle
    def convertToAckermann(self, kappa):
        return -math.atan(self.wheel_base * kappa)
 
    # this can be used to set the current pose
    def updatePos(self, pose):
        self.current_pose = pose
    
    # this can be used to set the new waypoint
    def updateWp(self, wp):
        self.next_wp[0] = wp[0] + self.offset
        self.next_wp[1] = wp[1]
        self.next_wp[2] = wp[2]

    # this calculates the curvature (pure pursuit algorithm)
    def calcCurvature(self, dist, y):
        if y == 0 or dist == 0:
            curvature = 1 / self.max_radius
        else:
            curvature = 2 * y / dist / dist            
            if abs(curvature) < 1 / self.max_radius:
                curvature = 1 / self.max_radius
        return curvature

    # this calculates the current distance from pose to waypoint
    def calcDistance(self):
        x = self.next_wp[0] - self.current_pos[0]
        y = self.next_wp[1] - self.current_pos[1]
        dst = math.sqrt(x*x + y*y)    
        return dst, y

    # this is the pipeline for calculating the steering angle
    # make sure, that current pose and waypoint is set before
    # calling this function
    def calcSteeringAngle(self):
        (dst, y) = self.calcDistance()
        kappa = self.calcCurvature(dst, y)
        self.steer = self.convertToAckermann(kappa)
        return dst

    # this calculates the gas and brake pedal positions
    # based on current velocity and dyn_thresh parameter
    # a threshold is calculated and compared to
    # v^2 - v0^2 = 2ax => a = (v^2 - v0^2) / (2x)
    # based on the result, the vehicle accelerates or brakes
    def calcPedalPositions(self, current_velo, dst):
        a = (self.wp_velocity * self.wp_velocity - current_velo * current_velo) / (2.0 * dst)
        if (a < 0.0):
            # brake
            self.gas = 0.0
            self.brake = abs(a / self.max_decel)
            if self.brake > 1.0:
                self.brake = 1.0
        else:
            # accelerate
            self.gas = abs(a / self.max_accel)
            self.brake = 0.0
            if self.gas > 1.0:
                self.gas = 1.0

    # use this function to calculate control commands
    def calcControlCommands(self, current_velo, waypoint):
        self.updateWp(waypoint)        
        dst = self.calcSteeringAngle()
        self.calcPedalPositions(current_velo, dst)
        return (round(self.steer, 4), round(self.brake, 4), round(self.gas, 4))


Linux/LaneDetector/lane_detector.py
# ======================================================
# Copyright (C) 2020 repa1030
# This program and the accompanying materials
# are made available under the terms of the MIT license.
# ======================================================

import numpy as np
import cv2

class SimpleLaneDetector:
    
    # constructor of class
    def __init__(self, src_points, dst_points, warp_img_size, look_ahead_ratio,
                    max_look_ahead_dst, pixel_to_cm, use_quadratic_interpolation,
                    draw_line_prediction, draw_waypoint, color_right_line, 
                    color_left_line, color_predicted_line, color_waypoint, 
                    draw_thickness, waypoint_radius):
        self.lane_width = 0.0
        self.i = 0
        self.warp_mat = cv2.getPerspectiveTransform(src_points, dst_points)
        self.warp_mat_inv = cv2.getPerspectiveTransform(dst_points, src_points)
        self.warp_size = warp_img_size
        self.la_ratio = look_ahead_ratio
        self.max_la = max_look_ahead_dst
        self.px2cm = pixel_to_cm
        self.use_quadratic = use_quadratic_interpolation
        self.line_prediction = draw_line_prediction
        self.draw_wp = draw_waypoint
        self.color_r = color_right_line
        self.color_l = color_left_line
        self.color_p = color_predicted_line
        self.color_wp = color_waypoint
        self.draw_thick = draw_thickness
        self.wp_radius = waypoint_radius

    # this is for drawing quadratic interpolation (unstable)
    def drawPolynom(self, img, detect, wp, min_y, max_y, width, height, poly_left, poly_right, dist=3):
        line_img = np.zeros_like(img).astype(np.uint8)
        i = int(min_y)
        if detect == 1:   
            while i > int(max_y):
                p1l = (int(poly_left(i)), int(i))
                p2l = (int(poly_left(i-dist)), int(i-dist))
                p1r = (int(poly_right(i)), int(i)) 
                p2r = (int(poly_right(i-dist)), int(i-dist))
                cv2.line(line_img, p1l, p2l, self.color_l, self.draw_thick)
                cv2.line(line_img, p1r, p2r, self.color_r, self.draw_thick)
                i -= dist
        elif detect == 2:
            while i > int(max_y):
                p1l = (int(poly_left(i)), int(i))
                p2l = (int(poly_left(i-dist)), int(i-dist))
                cv2.line(line_img, p1l, p2l, self.color_l, self.draw_thick)
                i -= dist
        elif detect == 3:
            while i > int(max_y):
                p1r = (int(poly_right(i)), int(i)) 
                p2r = (int(poly_right(i-dist)), int(i-dist))
                cv2.line(line_img, p1r, p2r, self.color_r, self.draw_thick)
                i -= dist
        if self.draw_wp:
            cv2.circle(line_img, (int(width/2 - (wp[1] * self.px2cm * 100.0)), int(height - (wp[0] * self.px2cm * 100.0))), self.wp_radius, self.color_wp, -1)
        return line_img

    # this function draws the final lane lines on the image
    def drawLines(self, img, lines, height, width, wp):
        line_img = np.zeros_like(img).astype(np.uint8)
        for line in lines:
            for x1, y1, x2, y2, color in line:
                cv2.line(line_img, (x1, y1), (x2, y2), color, self.draw_thick)
        if self.draw_wp:
            cv2.circle(line_img, (int(width/2.0 - (wp[1] * self.px2cm * 100.0)), int(height - (wp[0] * self.px2cm * 100.0))), self.wp_radius, self.color_wp, -1)
        return line_img

    # this function combines line and original image
    def combineImages(self, line_img, orig_img, width, height):
        line_img = cv2.warpPerspective(line_img, self.warp_mat_inv, (width, height))
        result = cv2.addWeighted(orig_img, 1, line_img, 0.8, 0.0)
        return result

    # this function calculates the next waypoint
    def calcWaypoint(self, x_left, x_right, y, width):
        wp = [0, 0, 0]
        # changing from openCV coord system to sensor coord system
        wp[0] = y
        wp[1] = (width - x_right - x_left) / 2.0 / self.px2cm / 100.0
        self.lane_width = x_right - x_left 
        return wp

    # this function calculates the next waypoint if only one lane is detected
    def calcWaypointWithOneLine(self, x, y, lane, width, det_lane):
        wp = [0, 0, 0]
        # left line is known
        if det_lane == 2:
            x_th = x + lane
            wp[1] = (width - x - x_th) / 2.0 / self.px2cm / 100.0
        #right line is known
        else:
            x_th = x - lane
            wp[1] = (width - x_th - x) / 2.0 / self.px2cm / 100.0
        wp[0] = y
        return wp

    # this is the pipeline for the lane detection
    def detectLaneLines(self, img_orig, velo):
        # waypoint (x=longitudinal, y=lateral, z=0) [m]
        wp = [0, 0, 0]
        # 0 -> no lines detected, 
        # 1 -> both lines detected,
        # 2 -> left line detected
        # 3 -> right line detected
        det_lines = 0
        height_orig = img_orig.shape[0]
        width_orig = img_orig.shape[1]
        warped = cv2.warpPerspective(img_orig, self.warp_mat, self.warp_size)
        height = warped.shape[0]
        width = warped.shape[1]
        
        gray_image = cv2.cvtColor(warped, cv2.COLOR_RGB2GRAY) 
        cannyed_image = cv2.Canny(gray_image, 100, 200)
        lines = cv2.HoughLinesP(
            cannyed_image,
            rho=6,
            theta=np.pi / 60,
            threshold=160,
            lines=np.array([]),
            minLineLength=40,
            maxLineGap=25
        )

        left_line_x = []
        left_line_y = []
        right_line_x = []
        right_line_y = []

        # Check if there are any hough lines
        if lines is None:
            return img_orig, wp

        # Add hough lines to right or left line
        for line in lines:
            for x1, y1, x2, y2 in line:
                if x1 < (width/2) and x2 < (width/2):
                    left_line_x.extend([x1, x2])
                    left_line_y.extend([y1, y2])
                elif x1 > (width/2) and x2 > (width/2):
                    right_line_x.extend([x1, x2])
                    right_line_y.extend([y1, y2])
                else:
                    continue

        min_y = int(height)
        max_y = int(height - (self.max_la * self.px2cm * 100))
        if max_y < 0:
            max_y = 0

        # Get left line polynom
        if not left_line_x or not left_line_y:
            left_line_det = False
        else:
            if self.use_quadratic:
                poly_left = np.poly1d(np.polyfit(
                    left_line_y,
                    left_line_x,
                    deg=2
                ))
            else:
                poly_left = np.poly1d(np.polyfit(
                    left_line_y,
                    left_line_x,
                    deg=1
                ))
                left_x_start = int(poly_left(max_y))
                left_x_end = int(poly_left(min_y))
            left_line_det = True

        # Get right line polynom
        if not right_line_x or not right_line_y:
            right_line_det = False
        else:
            if self.use_quadratic:
                poly_right = np.poly1d(np.polyfit(
                    right_line_y,
                    right_line_x,
                   deg=2
                ))
            else:
                poly_right = np.poly1d(np.polyfit(
                    right_line_y,
                    right_line_x,
                   deg=1
                ))
                right_x_start = int(poly_right(max_y))
                right_x_end = int(poly_right(min_y))
            right_line_det = True
        
        # calculate look ahead distance
        la_dst = velo * self.la_ratio
        if la_dst > self.max_la:
            la_dst = self.max_la
        la_x = height - (la_dst * 100.0 * self.px2cm)

        # Check how many lines are detected and calculate waypoint
        if left_line_det and right_line_det:
            if not self.use_quadratic:            
                available_lines = [
                    [[left_x_start, max_y, left_x_end, min_y, self.color_l]],
                    [[right_x_start, max_y, right_x_end, min_y, self.color_r]]
                    ]
            det_lines = 1
            wp = self.calcWaypoint(int(poly_left(la_x)), int(poly_right(la_x)), la_dst, width)
        elif left_line_det:
            poly_right = []
            det_lines = 2
            if not self.use_quadratic:
                if self.line_prediction and self.lane_width > 0.0:
                    available_lines = [
                        [[left_x_start, max_y, left_x_end, min_y, self.color_l]],
                        [[int(left_x_start + self.lane_width), max_y, int(left_x_end + self.lane_width), min_y, self.color_p]]
                        ]
                else:
                    available_lines = [
                        [[left_x_start, max_y, left_x_end, min_y, self.color_l]]
                        ]
            if self.lane_width > 0.0:
                wp = self.calcWaypointWithOneLine(int(poly_left(la_x)), la_dst, self.lane_width, width, det_lines)
        elif right_line_det:
            poly_left= []
            det_lines = 3
            if not self.use_quadratic:
                if self.line_prediction and self.lane_width > 0.0:
                    available_lines = [
                        [[int(right_x_start - self.lane_width), max_y, int(right_x_end - self.lane_width), min_y, self.color_p]],
                        [[right_x_start, max_y, right_x_end, min_y, self.color_r]]
                        ]
                else:
                    available_lines = [
                        [[right_x_start, max_y, right_x_end, min_y, self.color_r]]
                        ]
            if self.lane_width > 0.0:
                wp = self.calcWaypointWithOneLine(int(poly_right(la_x)), la_dst, self.lane_width, width, det_lines)
        else:
            poly_right = []
            poly_left = []
            if not self.use_quadratic:
                available_lines = []

        # draw lines on empty line image
        if self.use_quadratic:
            line_image = self.drawPolynom(
                warped, det_lines, wp, 
                min_y, max_y, 
                width, height,
                poly_left, poly_right
            )
        else:
            line_image = self.drawLines(
                warped, available_lines,
                height,
                width,
                wp
            )

        # warp line image and combine it with original image        
        output_image = self.combineImages(line_image, img_orig, width_orig, height_orig)

        # Return result
        return output_image, wp


Linux/lkas.py
# ======================================================
# Copyright (C) 2020 repa1030
# This program and the accompanying materials
# are made available under the terms of the MIT license.
# ======================================================

from TcpServer import tcp_server
from PurePursuit import pure_pursuit
from LaneDetector import lane_detector
import numpy as np
import cv2

if __name__=="__main__":

    ####################################################
    ################# Parameter Section ################
    ####################################################

    # General Settings
    display_detection = True        # Default: True
    display_warnings = False        # Default: False

    # TCP/IP Settings
    tcp_ip = '0.0.0.0'              # Default: '0.0.0.0'
    tcp_port = 5005                 # Default: 5005
    tcp_buffer_size = 1024          # Default: 1024 [bytes]
    tcp_time_out = 3.0              # Default: 3.0 [seconds]
    
    # Vehicle Settings
    command_velocity = 50.0         # Default: 50.0 [km/h]
    vehicle_mass = 1580.0           # Default: 1580.0 [kg]
    wheel_radius = 0.29             # Default: 0.29 [m]
    max_motor_torque = 1000.0       # Default: 1000.0 [Nm]
    max_brake_torque = 1500.0       # Default: 1500.0 [Nm]
    wheel_base = 2.44               # Default: 2.44 [m]
    
    # Pure Pursuit Settings
    look_ahead_ratio = 1.0          # Default: 1.0 (smoothing of steering)
    max_look_ahead_distance = 7.0   # Default: 7.0 [m] (if -1, it is calculated automatically)

    # CV Settings
    use_quadratic_interpol = False  # Default: False (don't use this)
    offset_to_clipping_plane = 3.83 # Default: 3.83 [m] (base link to bottom border of image) 
    pixel_to_cm = 0.25              # Default: 0.25 (change resolution of line image)
    padding_x = 150                 # Default: 150 [cm] (padding on both sides)
    padding_y = 700                 # Default: 700 [cm] (padding to forward)

    # Draw Settings
    draw_line_prediction = True     # Default: True (draw predicted line)
    draw_waypoint = True            # Default: True (draw next waypoint on image)
    draw_thickness = 3              # Default: 3 (thickness of lines and waypoint)
    waypoint_radius = 4             # Default: 4 (radius of the drawn waypoint)
    color_right_line = [0, 255, 0]  # Default: [0, 255, 0] [B G R]
    color_left_line = [255, 0, 0]   # Default: [255, 0, 0] [B G R]
    color_pred_line = [0, 0, 255]   # Default: [0, 0, 255] [B G R]
    color_waypoint = [0, 140, 255]  # Default: [0, 140, 255] [B G R]
    
    ####################################################
    ############# End of Parameter Section #############
    ####################################################

    src_p = np.float32([[150, 540], [808, 540], [571, 344], [389, 344]]) # BottomL, BottomR, TopR, TopL
    dst_p = np.float32([[int((0 + padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm)],
                        [int((300 + padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm)], 
                        [int((300 + padding_x) * pixel_to_cm), int((0 + padding_y) * pixel_to_cm)], 
                        [int((0 + padding_x) * pixel_to_cm), int((0 + padding_y) * pixel_to_cm)]])
    warp_size = (int((300 + 2 * padding_x) * pixel_to_cm), int((1000 + padding_y) * pixel_to_cm))
    if max_look_ahead_distance is -1:
        max_look_ahead_distance = (1000 + padding_y - 100)
    ld = lane_detector.SimpleLaneDetector(src_p, dst_p, warp_size, look_ahead_ratio,
                                            max_look_ahead_distance, pixel_to_cm, 
                                            use_quadratic_interpol, draw_line_prediction,
                                            draw_waypoint, color_right_line, color_left_line, 
                                            color_pred_line, color_waypoint, draw_thickness,
                                            waypoint_radius)
    print ('Initialized Instance of Lane Detector.')
    pp = pure_pursuit.PurePursuit(command_velocity, vehicle_mass, wheel_radius, 
                                    max_motor_torque, max_brake_torque, wheel_base,
                                    offset_to_clipping_plane)
    print ('Initialized Instance of Pure Pursuit.')
    tcp = tcp_server.TcpServer(tcp_ip, tcp_port, tcp_buffer_size, tcp_time_out)
    print ('Initialized Instance of TCP Server.')
    print ('Waiting for Connection...')
    tcp.connectToClient()
    connection = True
    print ('Connection to Client is established.')
    is_info_displayed = False
    
    while True:
        # If error occured -> try reconnecting
        if (connection is False):
            tcp.reconnect()
        # Receive message containing the image and the current velocity
        (image, current_velocity, connection) = tcp.receiveMessage()
        # Check if the image and the velocity are valid 
        if (image is None) or (current_velocity is None):
            if display_warnings:
                print ('Warning: Image transfer failed.')
            connection = tcp.sendMessage(0.0, 0.0, 0.0)
            continue
        # Display info that lkas is running
        if not is_info_displayed:
            print ('Lane Following Programm is running.')
            is_info_displayed = True
        current_velocity = current_velocity * 1000.0 / (60.0 * 60.0)
        # Detect the lane and calculate the waypoint
        (line_image, waypoint) = ld.detectLaneLines(image, current_velocity)
        if display_detection:
            cv2.imshow('Detection', line_image)
            cv2.waitKey(10)
        # Calculate the control commands
        (steerCmd, brakeCmd, gasCmd) = pp.calcControlCommands(current_velocity, waypoint)
        # Send the control commands back to Unity
        connection = tcp.sendMessage(steerCmd, brakeCmd, gasCmd)
    
    tcp.closeConnection()
        
